sydent/db/migration.py

```diff
@@ -0,0 +1,182 @@
+import json
+import sqlite3
+
+import signedjson.sign
+
+from sydent.util import json_decoder
+from sydent.util.emailutils import sendEmail
+from sydent.util.hash import sha256_and_url_safe_base64
+
+
+def update_assosc(self, conn: sqlite3.Connection):
+    # from table local_threepid_associations, select all associations where the medium is 'email'
+    cur = conn.cursor()
+
+    res = cur.execute(
+        "SELECT address, mxid FROM local_threepid_associations WHERE medium = 'email'"
+        "ORDER BY ts DESC"
+    )
+
+    associations: Dict[str, List[Tuple[str, str, str]]] = {}
+
+    # iterate through selected associations, casefold email, rehash it, and add to
+    # associations dict
+    for row in res.fetchall():
+        casefold_address = row[0].casefold()
+
+        # rehash the email since hash functions are case-sensitive
+        pepper_result = cur.execute("SELECT lookup_pepper from hashing_metadata")
+        pepper = pepper_result.fetchone()[0]
+        combo = "%s %s %s" % (casefold_address, "email", pepper)
+        lookup_hash = sha256_and_url_safe_base64(combo)
+
+        if casefold_address in associations:
+            associations[casefold_address].append((row[0], row[1], lookup_hash))
+        else:
+            associations[casefold_address] = [(row[0], row[1], lookup_hash)]
+
+    db_update_args: List[Tuple[str, str, str, str]] = []
+    to_delete: List[str] = []
+    mxids: List[str] = []
+
+    for casefold_address, assoc_tuples in associations.items():
+        db_update_args.append(
+            (
+                casefold_address,
+                assoc_tuples[0][2],
+                assoc_tuples[0][0],
+                assoc_tuples[0][1],
+            )
+        )
+
+        if len(assoc_tuples) > 1:
+            # Iterate over all associations except for the first one, since we've already
+            # processed it.
+            for assoc_tuple in assoc_tuples[1:]:
+                to_delete.append([assoc_tuple[0]])
+                mxids.append(assoc_tuple[1])
+
+    # iterate through the mxids and send email
+    for mxid in mxids:
+        templateFile = self.sydent.get_branded_template(
+            "matrix-org",
+            "migration_template.eml",
+            ("email", "email.template"),
+        )
+        res = cur.execute(
+            "SELECT address FROM local_threepid_associations WHERE mxid = ?", (mxid,)
+        )
+        address = res.fetchone()[0]
+        sendEmail(
+            self.sydent,
+            templateFile,
+            address,
+            {"mxid": "mxid", "subject_header_value": "MatrixID Deletion"},
+        )
+
+    if len(to_delete) > 0:
+        cur.executemany(
+            "DELETE FROM local_threepid_associations WHERE address = ?", to_delete
+        )
```

Comment:
I did a little experiment and it seems that `executemany` doesn't do anything if passed an empty list, but I'm a little torn because I think the `if` statement makes the code more clear. Happy to remove it if you feel strongly about it, but maybe it helps readability?

Reply:
sure, no worries. It at least avoids the question of whether anything will happen :)